design/Exceptions.md

```diff
@@ -0,0 +1,75 @@
+The behavior of every error case in WebGPU is defined by the spec on a
+case-by-case basis. A given error has one of these behaviors:
+
+* Synchronously throws a JS exception.
+* Occurs asynchronously - one of:
+  * Causes a WebGPU object to become internally null. Produces an error log entry.
+  * Causes an operation to no-op. Produces an error log entry.
+* If the operation returns a Promise, it rejects (and maybe produces an error log entry).
+
+(If a "developer mode" is enabled, all validation errors are thrown
+synchronously, as exceptions. Device loss may or may not be synchronous and
+this behavior may be implementation-specific.
+Out-of-memory errors should NOT be made synchronous if the application would
+otherwise have an opportunity to recover from them.)
+
+**The guidelines below are meant to help choose the individual cases defined by
+the spec, but every case must be specced. This does not allow for
+"implementation-defined" behavior.** Note that an implementation can easily
+surface a synchronous error "as-if" it's asynchronous, but it cannot do the
+opposite, so we prefer to err on the side of asynchronicity in the spec.
+
+As a general rule, those error cases should follow the following guidelines,
+but are allowed to deviate in individual cases. For WebGPU function call
+`o.f(a, b, ...)`, let `A = {a: a, b: b, ...}` represent the object graph
+passed into `o.f`.
+
+* If WebIDL's binding rules would throw an exception: Error **must** be synchronous.
+    * This should cover type checks.
```

Comment:
can we provide more detail here? will this cover function calls, field access, and enum values, or just the first two?

Reply:
Left intentionally vague because I don't know the exact semantics defined by WebIDL. But in general:
* accessing a unbound variable/global object throws an exception
* accessing a non-existent member (method, constant, attribute, property, etc) of an object gives you `undefined`.
* passing invalid types (as defined by WebIDL) into bindings tries to do a coercion, and throws an exception if it can't coerce.

Reply:
Updated as:

* If WebIDL's binding rules would throw an exception: Error **must** be synchronous.
    E.g.:
    * If a parameter is passed in which doesn't match the type declared by the WebIDL.